python/traverse.py

```diff
@@ -0,0 +1,179 @@
+from collections import deque
+from typing import Callable, Deque, List, Optional, Tuple, Union
+import timeit
+
+import lxml.html
+
+ElementReplacement = Optional[
+    Union[str, lxml.html.HtmlElement, List[lxml.html.HtmlElement]]
+]
+
+
+def serialize_str_or_element(e) -> str:
+    if isinstance(e, str):
+        return e
+    return lxml.html.tostring(e, encoding="unicode")
+
+
+def append_text(element: lxml.html.HtmlElement, text: str) -> None:
+    stripped_text = text.lstrip()
+
+    if element.text:
+        # print("appending element text", text)
+        stripped_element_text = element.text.rstrip()
+        sep = " " if stripped_element_text and stripped_text else ""
+        element.text = stripped_element_text + sep + stripped_text
+    else:
+        # print("setting element text", text)
+        element.text = stripped_text
+
+
+def append_tail(element: lxml.html.HtmlElement, text: str) -> None:
+    stripped_text = text.lstrip()
+
+    if element.tail:
+        # print("appending element tail", text)
+        stripped_element_tail = element.tail.rstrip()
+        sep = " " if stripped_element_tail and stripped_text else ""
+        element.tail = stripped_element_tail + sep + stripped_text
+    else:
+        # print("setting element tail", text)
+        element.tail = " " + stripped_text
+
+
+def traverse_and_execute(
+    html: str, fn: Callable[[lxml.html.HtmlElement], None]
+) -> list:
+    elements = lxml.html.fragments_fromstring(html)
+
+    # If there's leading text, the first element of the array will be a string.
+    # We can just discard that.
+    if isinstance(elements[0], str):
+        del elements[0]
+
+    for element in elements:
+        for e in element.iter():
+            fn(e)
+
+
+def get_source_definition(element: lxml.html.HtmlElement) -> str:
+    # https://html.spec.whatwg.org/multipage/parsing.html#escapingString
+    attributes = (
+        f'''{k}="{v.replace('"', "&quot;")}"''' for k, v in element.attrib.items()
+    )
+    return f"<{' '.join((element.tag, *attributes))}>"
+
+
+def traverse_and_replace(
+    html: str, replace: Callable[[lxml.html.HtmlElement], ElementReplacement]
+) -> str:
+    # Initialize result and work data structures
+    result: Deque[Union[str, lxml.html.HtmlElement]] = deque()
```

Comment:
My mistake on this one:
```suggestion
    result: Deque[str] = deque()
```